src/libstd/io/buffered.rs

```
@@ -0,0 +1,637 @@
+    }
+
+    /// Creates a new `BufWriter` with the specified buffer capacity
+    pub fn with_capacity(cap: usize, inner: W) -> BufWriter<W> {
+        BufWriter {
+            inner: Some(inner),
+            buf: Vec::with_capacity(cap),
+        }
+    }
+
+    fn flush_buf(&mut self) -> io::Result<usize> {
+        let len = self.buf.len();
+        let n = try!(self.inner.as_mut().unwrap().write(&self.buf[]));
+        // FIXME would be better expressed as .remove(0..n) if it existed
+        unsafe {
+            ptr::copy_memory(self.buf.as_mut_ptr(),
+                             self.buf.as_ptr().offset(n as isize),
+                             len - n);
+        }
+        self.buf.truncate(len - n);
+        Ok(n)
+    }
+
+    /// Gets a reference to the underlying writer.
+    pub fn get_ref(&self) -> &W { self.inner.as_ref().unwrap() }
+
+    /// Gets a mutable reference to the underlying write.
+    ///
+    /// # Warning
+    ///
+    /// It is inadvisable to directly read from the underlying writer.
+    pub fn get_mut(&mut self) -> &mut W { self.inner.as_mut().unwrap() }
+
+    /// Unwraps this `BufWriter`, returning the underlying writer.
+    ///
+    /// The buffer is flushed before returning the writer.
+    pub fn into_inner(mut self) -> W {
+        // FIXME(#12628): is panicking the right thing to do if flushing panicks?
+        self.flush_buf().unwrap();
+        self.inner.take().unwrap()
+    }
+}
+
+impl<W: Write> Write for BufWriter<W> {
+    fn write(&mut self, buf: &[u8]) -> io::Result<usize> {
+        if self.buf.len() + buf.len() > self.buf.capacity() {
+            try!(self.flush_buf());
+        }
+        if self.buf.len() == 0 && buf.len() >= self.buf.capacity() {
+            self.inner.as_mut().unwrap().write(buf)
+        } else {
+            let amt = cmp::min(buf.len(), self.buf.capacity());
+            Write::write(&mut self.buf, &buf[..amt])
+        }
+    }
+
+    fn flush(&mut self) -> io::Result<()> {
```

flush should probably return an error if flush_buf doesn't write all bytes.